Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -0,0 +1,423 @@
+    compose_rightPart_eq_rightPart (isBalanced_treeEquivInvFun' _)]
+  exact ⟨ttl, ttr⟩
+
+/-- Equivalence between Dyck words and rooted binary trees. -/
+def treeEquiv : { p : DyckPath // p.IsBalanced } ≃ Tree Unit where
+  toFun := treeEquivToFun
+  invFun := treeEquivInvFun
+  left_inv st := treeEquiv_left_inv st rfl
+  right_inv := treeEquiv_right_inv
+
+@[nolint unusedHavesSuffices]
+theorem length_eq_two_mul_iff_numNodes_eq {n : ℕ} (st : { p : DyckPath // p.IsBalanced }) :
+    st.1.length = 2 * n ↔ (treeEquiv st).numNodes = n := by
+  induction' n using Nat.strongInductionOn with n ih generalizing st
+  cases' eq_or_ne st.1 [] with j j
+  · rw [j, length_nil, zero_eq_mul, treeEquiv, Equiv.coe_fn_mk, treeEquivToFun]
+    simp_rw [j, dite_true, numNodes]; tauto
+  rw [treeEquiv, Equiv.coe_fn_mk, treeEquivToFun]
+  simp_rw [j, dite_false, numNodes]
+  obtain ⟨p, bp⟩ := st
+  dsimp only at j ⊢
+  have bl := leftPart_isBalanced bp j
+  have br := rightPart_isBalanced bp j
+  obtain ⟨sl, le⟩ := bl.length_even
+  obtain ⟨sr, re⟩ := br.length_even
+  rw [← two_mul] at le re
+  conv_lhs =>
+    rw [eq_U_leftPart_D_rightPart bp j, length_append, length_append, length_append,
+      length_singleton, length_singleton, le, re,
+      show 1 + 2 * sl + 1 + 2 * sr = 2 * (sl + sr + 1) by omega]
+  constructor <;> intro h
+  · have tl := (ih sl (by omega) ⟨_, bl⟩).mp le
+    have tr := (ih sr (by omega) ⟨_, br⟩).mp re
+    dsimp only [treeEquiv, Equiv.coe_fn_mk] at tl tr
+    rw [tl, tr]; exact mul_left_cancel₀ two_ne_zero h
+  · set wl := Subtype.mk _ bl
+    set wr := Subtype.mk _ br
+    have ntwl_lt : (treeEquivToFun wl).numNodes < n := by omega
+    have ntwr_lt : (treeEquivToFun wr).numNodes < n := by omega
+    have el := ih _ ntwl_lt ⟨_, bl⟩
+    have er := ih _ ntwr_lt ⟨_, br⟩
+    simp only [treeEquiv, Equiv.coe_fn_mk, iff_true] at el er
+    rw [le, Nat.mul_left_cancel_iff zero_lt_two] at el
+    rw [re, Nat.mul_left_cancel_iff zero_lt_two] at er
+    rw [← el, ← er] at h; omega
+
+/-- Equivalence between Dyck words of length `2 * n` and
+rooted binary trees with `n` internal nodes. -/
+def finiteTreeEquiv (n : ℕ) :
+    { p : DyckPath // p.IsBalanced ∧ p.length = 2 * n } ≃ treesOfNumNodesEq n where
+  toFun := fun ⟨p, ⟨bp, len⟩⟩ ↦
+    ⟨treeEquiv ⟨p, bp⟩, by rw [mem_treesOfNumNodesEq, ← length_eq_two_mul_iff_numNodes_eq, len]⟩
+  invFun := fun ⟨tr, sz⟩ ↦
+    ⟨(treeEquiv.symm tr).1, ⟨(treeEquiv.symm tr).2, by rwa [length_eq_two_mul_iff_numNodes_eq,
+      ← mem_treesOfNumNodesEq, Equiv.apply_symm_apply]⟩⟩
+  left_inv := fun ⟨p, ⟨bp, len⟩⟩ ↦ by simp only [Equiv.symm_apply_apply]
+  right_inv := fun ⟨tr, sz⟩ ↦ by simp only [Subtype.coe_eta, Equiv.apply_symm_apply]
+
+instance instFintypeDyckWords {n : ℕ} :
+    Fintype { p : DyckPath // p.IsBalanced ∧ p.length = 2 * n } :=
+  Fintype.ofEquiv _ (finiteTreeEquiv n).symm
+
+/-- There are `catalan n` Dyck words of length `2 * n`. -/
+theorem card_dyckPath_isBalanced_eq_catalan (n : ℕ) :
```

Comment:
I had overlooked you already were in the `DyckPath` namespace, so this is a more concise name
```suggestion
theorem card_isBalanced_eq_catalan (n : ℕ) :
```